File: backend/app/routers/expenses.py
"""花費記錄（SRS 4.5）。路由只做參數接收與回應，業務規則在 services/expenses.py。

- GET    /expenses?start_date&end_date   花費列表（消費日區間，兩端皆含、皆可省略）／400
- POST   /expenses                       201／400／409（所屬月份已結算）
- PUT    /expenses/{id}                  200／400／404／409
- DELETE /expenses/{id}                  204／404／409
"""

from fastapi import APIRouter, Response, status

from app.deps import CurrentUserId, DbSession
from app.schemas.common import error_responses
from app.schemas.expense import ExpenseCreate, ExpenseOut, ExpenseUpdate
from app.services import expenses as service

router = APIRouter(prefix="/expenses", tags=["expenses"])


@router.get("", response_model=list[ExpenseOut], responses=error_responses(400))
def list_expenses(
    db: DbSession,
    user_id: CurrentUserId,
    start_date: str | None = None,
    end_date: str | None = None,
) -> list[ExpenseOut]:
    return [ExpenseOut.model_validate(e) for e in service.list_expenses(db, user_id, start_date, end_date)]


@router.post(
    "",
    response_model=ExpenseOut,
    status_code=status.HTTP_201_CREATED,
    responses=error_responses(400, 409),
)
def create_expense(payload: ExpenseCreate, db: DbSession, user_id: CurrentUserId) -> ExpenseOut:
    return ExpenseOut.model_validate(service.create_expense(db, user_id, payload))


@router.put("/{expense_id}", response_model=ExpenseOut, responses=error_responses(400, 404, 409))
def update_expense(
    expense_id: int, payload: ExpenseUpdate, db: DbSession, user_id: CurrentUserId
) -> ExpenseOut:
    return ExpenseOut.model_validate(service.update_expense(db, user_id, expense_id, payload))


@router.delete(
    "/{expense_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    responses=error_responses(404, 409),
)
def delete_expense(expense_id: int, db: DbSession, user_id: CurrentUserId) -> Response:
    service.delete_expense(db, user_id, expense_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/api/expenses.ts
/** 花費 API（SRS 4.5）。型別全部來自 schema.d.ts，不手寫。 */
import { apiFetch, resolvePath, withQuery } from "./client";
import type { components } from "./schema";

export type Expense = components["schemas"]["ExpenseOut"];
export type ExpenseCreate = components["schemas"]["ExpenseCreate"];
export type ExpenseUpdate = components["schemas"]["ExpenseUpdate"];
export type PaymentMethod = Expense["payment_method"];

/** GET /expenses?start_date&end_date（兩端皆含） */
export function listExpenses(startDate: string, endDate: string): Promise<Expense[]> {
  return apiFetch<Expense[]>(withQuery("/api/v1/expenses", { start_date: startDate, end_date: endDate }));
}

export function createExpense(payload: ExpenseCreate): Promise<Expense> {
  return apiFetch<Expense>("/api/v1/expenses", { method: "POST", json: payload });
}

export function updateExpense(id: number, payload: ExpenseUpdate): Promise<Expense> {
  return apiFetch<Expense>(resolvePath("/api/v1/expenses/{expense_id}", { expense_id: id }), { method: "PUT", json: payload });
}

export function deleteExpense(id: number): Promise<void> {
  return apiFetch<void>(resolvePath("/api/v1/expenses/{expense_id}", { expense_id: id }), { method: "DELETE" });
}
